frame/offences/src/lib.rs

```diff
@@ -106,53 +103,18 @@ decl_storage! {
 decl_event!(
 	pub enum Event {
 		/// There is an offence reported of the given `kind` happened at the `session_index` and
-		/// (kind-specific) time slot. This event is not deposited for duplicate slashes. last
-		/// element indicates of the offence was applied (true) or queued (false)
-		/// \[kind, timeslot, applied\].
-		Offence(Kind, OpaqueTimeSlot, bool),
+		/// (kind-specific) time slot. This event is not deposited for duplicate slashes.
+		/// \[kind, timeslot\].
+		Offence(Kind, OpaqueTimeSlot),
 	}
 );
 
 decl_module! {
 	pub struct Module<T: Config> for enum Call where origin: T::Origin {
 		fn deposit_event() = default;
 
-		fn on_initialize(now: T::BlockNumber) -> Weight {
-			// only decode storage if we can actually submit anything again.
-			if !T::OnOffenceHandler::can_report() {
-				return 0;
-			}
-
-			let limit = T::WeightSoftLimit::get();
-			let mut consumed = Weight::zero();
-
-			<DeferredOffences<T>>::mutate(|deferred| {
-				deferred.retain(|(offences, perbill, session)| {
-					if consumed >= limit {
-						true
-					} else {
-						// keep those that fail to be reported again. An error log is emitted here; this
-						// should not happen if staking's `can_report` is implemented properly.
-						match T::OnOffenceHandler::on_offence(&offences, &perbill, *session) {
-							Ok(weight) => {
-								consumed += weight;
-								false
-							},
-							Err(_) => {
-								log::error!(
-									target: "runtime::offences",
-									"re-submitting a deferred slash returned Err at {:?}. \
-									 This should not happen with pallet-staking",
-									now,
-								);
-								true
-							},
-						}
-					}
-				})
-			});
-
-			consumed
+		fn on_runtime_upgrade() -> Weight {
+			migration::remove_deferred_storage::<T>()
```

Comment:
Is it necessary to verify the storage version before call this method?

Reply:
Yes! otherwise, if we don't remove the code before the next release, you end up running this migration twice. In this case, it won't be a disaster, but you should make sure that the migration happens only once. 

You have two ways to do so: 

1. Introduce a new `Versions` storage item here, like `pallet-staking`.
2. Use the crate version, and move the crate to `4.0.0`. A bit more involved, like `pallet-elections-phragmen`.

Reply:
Okay, seeing that in the first line of the migration you short circuit if the storage does not exist, then it might be okay to leave this as-is.

Reply:
Last suggestion: please change the name as such to show that this function is not repeatedly calling a migration. 

```suggestion
			migration::maybe_remove_deferred_storage::<T>()
```